dbms/src/Interpreters/sortBlock.cpp

```diff
@@ -116,38 +122,37 @@ void sortBlock(Block & block, const SortDescription & description, UInt64 limit)
             ? block.getByName(description[0].column_name).column.get()
             : block.safeGetByPosition(description[0].column_number).column.get();
 
-        IColumn::Permutation perm;
-        if (auto column_string_ptr = needCollation(column, description[0]))
+        if (isCollationRequired(description[0]))
         {
-            const ColumnString & column_string = typeid_cast<const ColumnString &>(*column_string_ptr);
-            column_string.getPermutationWithCollation(*description[0].collator, reverse, limit, perm);
+            /// it it's real string column, than we need sort
+            if (const ColumnString * column_string = checkAndGetColumn<ColumnString>(column))
+                column_string->getPermutationWithCollation(*description[0].collator, reverse, limit, perm);
+            else if (!checkAndGetColumnConstData<ColumnString>(column))
+                throw Exception("Collations could be specified only for String columns.", ErrorCodes::BAD_COLLATION);
+
         }
-        else
+        else if (!isColumnConst(*column))
             column->getPermutation(reverse, limit, description[0].nulls_direction, perm);
+        /// we don't need to do anything with const column
 
         size_t columns = block.columns();
         for (size_t i = 0; i < columns; ++i)
             block.getByPosition(i).column = block.getByPosition(i).column->permute(perm, limit);
     }
     else
     {
-        size_t size = block.rows();
-        IColumn::Permutation perm(size);
-        for (size_t i = 0; i < size; ++i)
```

Comment:
Do not understand why this code was moved up.
Now, if `description.size() == 1`, we do it twice: here and in `ColumnString::getPermutationWithCollation`